### ISSUE TITLE
utils.py

### ISSUE THREAD

**Issue by karthik78437** [karthik78437/karthik#5]
"""
utils.py

Utility helpers: train/test split, scaling, metrics (RMSE, MAE, MAPE), saving & loading.
"""

import numpy as np
from sklearn.preprocessing import StandardScaler
import joblib
import os
from sklearn.metrics import mean_squared_error, mean_absolute_error

def train_test_split_sequences(X, y, test_size=0.2, shuffle=False):
    """
    Split sequences while preserving time order.
    """
    n = len(X)
    split = int(n * (1 - test_size))
    X_train = X[:split]
    y_train = y[:split]
    X_test = X[split:]
    y_test = y[split:]
    return X_train, X_test, y_train, y_test

def scale_sequences(X_train, X_test, scaler_path=None):
    """
    Scale features per feature column using StandardScaler.
    Arguments:
        X_train: (n_samples, seq_len, n_feats)
    Returns:
        X_train_scaled, X_test_scaled, scaler (fitted on flattened training features)
    """
    n_samples, seq_len, n_feats = X_train.shape
    scaler = StandardScaler()
    # flatten time and sample for scaling
    X_train_flat = X_train.reshape(-1, n_feats)
    scaler.fit(X_train_flat)
    X_train_s = scaler.transform(X_train_flat).reshape(n_samples, seq_len, n_feats)
    X_test_flat = X_test.reshape(-1, n_feats)
    X_test_s = scaler.transform(X_test_flat).reshape(X_test.shape[0], seq_len, n_feats)
    if scaler_path:
        joblib.dump(scaler, scaler_path)
    return X_train_s, X_test_s, scaler

def rmse(true, pred):
    return np.sqrt(mean_squared_error(true, pred))

def mape(true, pred):
    # avoid divide by zero
    true, pred = np.array(true), np.array(pred)
    denom = np.where(np.abs(true) < 1e-6, 1e-6, true)
    return np.mean(np.abs((true - pred) / denom)) * 100.0

def save_model_h5(model, path):
    model.save(path)

def load_model_h5(path):
    from tensorflow.keras.models import load_model
    # if model includes custom layer, pass custom_objects
    return load_model(path, custom_objects={'TemporalAttention': __import__('attention_layer').attention_layer.TemporalAttention})
